notebooks/working/3.0-mab-independent-variable-simulation.md

```diff
@@ -34,487 +47,269 @@ import random
+            if isCategorical(var_type):
+                # If only one category
+                if isUnique(alt_data[alt_var]):
+                    # Add name of alternative to variable
+                    # and store distriburion & parameters
+                    alt_spec_var_dic = getConstantDistAlt(alt_var, alt_data, alt_name_dic, alt)
+                else:
+                    # If more than one category, compute
+                    # the frequency of values and store
+                    # as parameters
+                    # Add name of alternative to variable
+                    # and store distriburion & parameters
+                    alt_spec_var_dic = getCategoricalDistAlt(alt_var, alt_data, alt_name_dic, alt)
+            else:
+                # If data is not categorical but has one unique value
+                if isUnique(alt_data[alt_var]):
+                    alt_spec_var_dic = getConstantDistAlt(alt_var, alt_data, alt_name_dic, alt)
+                # If data is not categorical but has more than one unique value
+                else:
+                    alt_spec_var_dic = getContinuousDistAlt(alt_var, alt_data, alt_name_dic, alt, cont_dists)
+            alt_spec_dic.update(alt_spec_var_dic)
+        all_alt_spec_var_dic.update(alt_spec_dic)
+    return all_alt_spec_var_dic
 
-```python
-# # Create a copy of the data subset
-# subset_bike_data = bike_data_long[model_variables].copy()
-```
+##############################################################
+##############################################################
 
-```python
-# # Create the needed variables for the conversion
-# ind_spec_vars = ['num_kids', 'household_size',
-#                  'cars_per_licensed_drivers', 'cross_bay',
-#                  'num_cars', 'num_licensed_drivers']
-
-# alt_spec_vars = ['total_travel_time', 'total_travel_distance']
-
-# subset_spec_vars = {'total_travel_cost': [1, 2, 3],
-#                     'cost_per_distance': [1, 2, 3]}
-
-# alternative_name_dict = {1: 'drive_alone',
-#                          2: 'shared_2',
-#                          3: 'shared_3p',
-#                          4: 'wtw',
-#                          5: 'dtw',
-#                          6: 'wtd',
-#                          7: 'walk',
-#                          8: 'bike'}
-```
+# Code for Trip Specific Variables
+# Loop around the different available alternatives
+def tripSpecDist(data_long, obs_id_col, trip_spec, cont_dists):
+    # Trip Specific Variable (maybe combine with individual specific variables)
+    # Loop around trip (observation) specific variables
+    trip_spec_dict = defaultdict(dict)
+    for trip_var in trip_spec:
+        # generate array of values for trip specific variable
+        trip_var_data = pd.Series([(data_long.loc[data_long[obs_id_col] == x][trip_var].unique()[0]) for x in data_long[obs_id_col].unique()])
+        # Get distribution if variable is categorical
+        var_type = var_types[trip_var]
```

Comment:
I believe you are relying on a global variable here (which you shouldn't do).
`var_types` is not defined as one of the function inputs, and it is not generated between lines 199 and 205.

Reply:
Good catch! totally missed that.